cmake fails to build with -DZSTD_PROGRAMS_LINK_SHARED=ON on windows

Hi,

building with  -DZSTD_PROGRAMS_LINK_SHARED=ON results in:
```
fileio_asyncio.c.obj : error LNK2019: unresolved external symbol POOL_create referenced in function AIO_IOPool_createThreadPool
fileio_asyncio.c.obj : error LNK2019: unresolved external symbol POOL_free referenced in function AIO_IOPool_destroy
fileio_asyncio.c.obj : error LNK2019: unresolved external symbol POOL_joinJobs referenced in function AIO_IOPool_join
fileio_asyncio.c.obj : error LNK2019: unresolved external symbol POOL_add referenced in function AIO_IOPool_enqueueJob
programs\zstd.exe : fatal error LNK1120: 4 unresolved externals
```

The reason is fairly simple: The API in pool.h isn't made visible, therefore the references from programs/fileio.c can't see them.

Before 1598e6c6 a build like this worked. CC: @yoniko 

It looks like it'd be easy enough to fix for POOL_create, POOL_free, there are public wrappers. But for the rest there isn't an obvious answer.

Regards,

Andres

> I think a similar approach to https://github.com/facebook/zstd/pull/3496 could be implemented for the cmake build.

That indeed works. Oddly enough, zstd-dll also has a custom xxhash.c reference, which doesn't seem to be required with msvc.

Ah, and I see why this doesn't fail on !windows: While lib/Makefile adds `-fvisibility=hidden`, build/cmake/lib/CMakeLists.txt doesn't.

If you add `set_target_properties(libzstd_shared PROPERTIES C_VISIBILITY_PRESET hidden)` it fails on linux as well:
```
/usr/bin/ld: programs/CMakeFiles/zstd.dir/home/andres/src/zstd/programs/fileio_asyncio.c.o: in function `AIO_IOPool_init':
fileio_asyncio.c:(.text+0xde): undefined reference to `POOL_create'
/usr/bin/ld: programs/CMakeFiles/zstd.dir/home/andres/src/zstd/programs/fileio_asyncio.c.o: in function `AIO_WritePool_enqueueAndReacquireWriteJob':
fileio_asyncio.c:(.text+0xf67): undefined reference to `POOL_add'
...
```

Which is nice, because it allows fixing this outside of windows :)

--- a/build/cmake/programs/CMakeLists.txt
+++ b/build/cmake/programs/CMakeLists.txt
@@ -32,7 +32,17 @@ if (MSVC)
     set(PlatformDependResources ${MSVC_RESOURCE_DIR}/zstd.rc)
 endif ()
 
-add_executable(zstd ${PROGRAMS_DIR}/zstdcli.c ${PROGRAMS_DIR}/util.c ${PROGRAMS_DIR}/timefn.c ${PROGRAMS_DIR}/fileio.c ${PROGRAMS_DIR}/fileio_asyncio.c ${PROGRAMS_DIR}/benchfn.c ${PROGRAMS_DIR}/benchzstd.c ${PROGRAMS_DIR}/datagen.c ${PROGRAMS_DIR}/dibio.c ${PROGRAMS_DIR}/zstdcli_trace.c ${PlatformDependResources})
+set(ZSTD_PROGRAM_SRCS ${PROGRAMS_DIR}/zstdcli.c ${PROGRAMS_DIR}/util.c
+    ${PROGRAMS_DIR}/timefn.c ${PROGRAMS_DIR}/fileio.c
+    ${PROGRAMS_DIR}/fileio_asyncio.c ${PROGRAMS_DIR}/benchfn.c
+    ${PROGRAMS_DIR}/benchzstd.c ${PROGRAMS_DIR}/datagen.c
+    ${PROGRAMS_DIR}/dibio.c ${PROGRAMS_DIR}/zstdcli_trace.c
+    ${PlatformDependResources})
+if (MSVC AND ZSTD_PROGRAMS_LINK_SHARED)
+    list(APPEND ZSTD_PROGRAM_SRCS ${LIBRARY_DIR}/common/pool.c ${LIBRARY_DIR}/common/threading.c)
+endif ()
+
+add_executable(zstd ${ZSTD_PROGRAM_SRCS})
 target_link_libraries(zstd ${PROGRAMS_ZSTD_LINK_TARGET})
 if (CMAKE_SYSTEM_NAME MATCHES "(Solaris|SunOS)")
     target_link_libraries(zstd rt)
@@ -75,7 +85,9 @@ if (UNIX)
         ${CMAKE_CURRENT_BINARY_DIR}/zstdless.1
         DESTINATION "${MAN_INSTALL_DIR}")
 
-    add_executable(zstd-frugal ${PROGRAMS_DIR}/zstdcli.c ${PROGRAMS_DIR}/util.c ${PROGRAMS_DIR}/timefn.c ${PROGRAMS_DIR}/fileio.c ${PROGRAMS_DIR}/fileio_asyncio.c)
+    add_executable(zstd-frugal ${PROGRAMS_DIR}/zstdcli.c
+        ${PROGRAMS_DIR}/util.c ${PROGRAMS_DIR}/timefn.c
+        ${PROGRAMS_DIR}/fileio.c ${PROGRAMS_DIR}/fileio_asyncio.c)
     target_link_libraries(zstd-frugal ${PROGRAMS_ZSTD_LINK_TARGET})
     set_property(TARGET zstd-frugal APPEND PROPERTY COMPILE_DEFINITIONS "ZSTD_NOBENCH;ZSTD_NODICT;ZSTD_NOTRACE")
 endif ()

